source/common/config/xds_manager_impl.h

```diff
@@ -56,6 +77,15 @@ class XdsManagerImpl : public XdsManager {
   // prior to the cluster-manager deletion.
   Upstream::ClusterManager* cm_;
   GrpcMuxSharedPtr ads_mux_;
+
+  // Stores all authorities as configured in the bootstrap.
+  // Also includes the default config-source (the one that will be used for cases where the
+  // authority in a resource doesn't exist, or doesn't match).
+  std::vector<AuthorityData> authorities_;
+
+  // The default authority that will be used for cases where the authority in a resource doesn't
+  // exist, or doesn't match.
+  AuthorityDataIterator default_authority_;
```

Comment:
It looks like the you can be adding more elements into `authorities_` after setting the `default_authority_`. Is it the case? If so the iterator might become dangling if vector reallocates the memory.

I think in general references to class members are problematic. I would even prefer to simply copy the element, instead of keeping an iterator to a member vector.

Reply:
The `addAuthority()` method is the only one that updates this vector.
The change to the [`initializeAdsConnections()` method](https://github.com/envoyproxy/envoy/pull/39781/files#diff-0412531dda7a6cff7a5b1f21b962b830df38d0b3a92cdacc56180ee1beb515e5R67-R79) adds all the config-sources in the following order:
1. all the authorities defined in [config_sources](https://github.com/envoyproxy/envoy/blob/4861b2057afbc67290da0ec77ae1f1af60f152d0/api/envoy/config/bootstrap/v3/bootstrap.proto#L348).
2. the authority from the [default_config_source](https://github.com/envoyproxy/envoy/blob/4861b2057afbc67290da0ec77ae1f1af60f152d0/api/envoy/config/bootstrap/v3/bootstrap.proto#L353C24-L353C45).

Authorities are not added dynamically, and are only defined in the bootstrap.

Reply:
Understood.

This is a brittle structure however. These two members are tightly bound and there is additional requirement that vector does not change after iterator is set. This requirement is not clear from the code. This is a bit of a nit, I would encapsulate the vector and iterator into a dedicated type given their relationship and then make it a std::unique_ptr<const XXX> member in this class.

Reply:
I've decided to split them to a vector of all the config sources, and a pointer to the default-config-source. I believe it addresses your concern (and keeps things simple), but let me know if you think otherwise.